### PR TITLE
[Text input] Add ability for text input to let platform can unfocus.

### PR DESCRIPTION
**Pull request by luckysmg**
*List which issues are fixed by this PR. You must list at least one issue.*
- https://github.com/flutter/flutter/issues/123523

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

**Review by justinmc — APPROVED** [review]
LGTM with one improvement to the docs.  Thank you for fixing this!

**Comment by goderbauer** [comment]
(triage fly-by): @justinmc looks like Google Testing is unhappy with this one.

**Comment by luckysmg** [comment]
Seems that `connectionClosed` can do this thing. I will have a check. If that functions works, I think I can close this PR and just use that function^_^

**Comment by luckysmg** [comment]
Tested that `connectionClosed` can do this thing.

**Comment by luckysmg** [comment]
Reopen.

Context: https://github.com/flutter/flutter/issues/123523#issuecomment-1489872260

**Comment by luckysmg** [comment]
@LongCatIsLooong would you mind share your thoughts on this ^_^ ？

**Comment by luckysmg** [comment]
Also not sure why Google testing failed.. I m not googler, @justinmc  would you mind help me on that ^_^ ?

File: packages/flutter/lib/src/services/text_input.dart [hunk 1]
```diff
@@ -1177,6 +1177,13 @@ mixin TextInputClient {
   /// `NSStandardKeyBindingResponding` protocol or user-specified selector
   /// from `DefaultKeyBinding.Dict`.
   void performSelector(String selectorName) {}
+
+  /// Request client to unfocus.
+  ///
+  /// For example, when closes text input from platform (eg: Some third party keyboard can close
+  /// keyboard from native keyboard view), and this method will be called to update the focus state
+  /// in framework.
```
[the remark this answers is not in the document]
Comment (luckysmg) [reply packages/flutter/lib/src/services/text_input.dart]: Done

**Comment by justinmc** [comment]
Ah, this is a breaking change for anyone that `implements` TextInputClient.  They will need to provide an implementation of `unfocus`.

Can you take a look at the process in the [breaking changes guide](https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes)?  I think we'll need to write a migration guide, even though it should be simple.

**Comment by luckysmg** [comment]
Emm why it is a break change? the text input client default unfocus implementation is empty.🤔

**Comment by justinmc** [comment]
Yeah exactly.  Probably people should be using `with` if they don't want to be broken when a method is added.

**Comment by luckysmg** [comment]
I m thinking to reuse the function `connectionClosed` without any break changes. Because the `connectionClosed` behavior maybe different on web and mobile. 

I will have that try. ^_^

**Comment by justinmc** [comment]
That sounds like it could work too.  Let me know how it goes.

**Comment by luckysmg** [comment]
I m going to land https://github.com/flutter/flutter/pull/123929 since we can use exist function to do things.

**Comment by luckysmg** [comment]
Mark as closed. ^_^

**Pull request closed without merging**